### PR TITLE
Refactor functions that extract a range of samples from TLS and TLV.

### PR DESCRIPTION
**Pull request by mzient**
Signed-off-by: Michał Zientkiewicz <mzient@gmail.com>

#### Why we need this PR?
*Pick one, remove the rest*
- Refactoring to improve consistency
- Fix bug in assert condition that disallowed empty results

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
      * Rename subshape to `sample_range`.
      * Reimplement `sample_range` for TensorListView with one for TensorListShape.
      * Add step to `sample_range` for TensorListView.
      * Fix boundary condition assertions
      * Add extended `ndim` handling (already present in TLV variant)
 - Affected modules and functionalities:
     * `sample_range`; renamed `subshape` to `sample_range` when necessary
 - Key points relevant for the review:
     * N/A
 - Validation and testing:
     * Unit tests
 - Documentation (including examples):
     * Doxygen

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by mzient** [comment]
!build

File: include/dali/core/tensor_shape.h [hunk 1]
```diff
@@ -1435,23 +1435,40 @@ auto permute_dims(const TensorListShape<in_ndim> &in, const Permutation &axis_or
...
-  assert(step >= 1);
-  assert(end > begin);
-  assert(begin >= 0);
-  auto div_ceil = [](int x, int y) { return 1 + ((x - 1) / y); };
+template <int out_ndim, int ndim>
+void sample_range(TensorListShape<out_ndim> &out, const TensorListShape<ndim> &in,
+                  int begin, int end, int step = 1) {
+  detail::check_compatible_ndim<out_ndim, ndim>();
+  assert(begin >=0 && begin <= in.num_samples());
+  assert(end >= begin && end <= in.num_samples());
```
Comment (szalpal) [inline include/dali/core/tensor_shape.h:1449]: Please add test for corner case `begin=0, end=0` (or `begin == end` in general)

**Review by szalpal — APPROVED** [review]

File: include/dali/core/tensor_shape.h [hunk 2]
```diff
@@ -1435,23 +1435,40 @@ auto permute_dims(const TensorListShape<in_ndim> &in, const Permutation &axis_or
...
-                                int step = 1) {
-  assert(step >= 1);
-  assert(end > begin);
-  assert(begin >= 0);
-  auto div_ceil = [](int x, int y) { return 1 + ((x - 1) / y); };
+template <int out_ndim, int ndim>
+void sample_range(TensorListShape<out_ndim> &out, const TensorListShape<ndim> &in,
+                  int begin, int end, int step = 1) {
+  detail::check_compatible_ndim<out_ndim, ndim>();
+  assert(begin >=0 && begin <= in.num_samples());
```
Comment (JanuszL) [inline include/dali/core/tensor_shape.h:1448]:
```suggestion
  assert(begin >= 0 && begin <= in.num_samples());
```

**Review by JanuszL — APPROVED** [review]

File: include/dali/core/tensor_shape.h (hunk 1 above)
Comment (mzient) [reply include/dali/core/tensor_shape.h:1449]: Done.

**Comment by mzient** [comment]
!build
